url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/Sub.java

```diff
@@ -0,0 +1,91 @@
+/*
+    Copyright (c) 2007-2012 iMatix Corporation
+    Copyright (c) 2009-2011 250bpm s.r.o.
+    Copyright (c) 2007-2011 Other contributors as noted in the AUTHORS file
+
+    This file is part of 0MQ.
+
+    0MQ is free software; you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    0MQ is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+package zmq;
+
+public class Sub extends XSub {
+
+    public static class SubSession extends XSub.XSubSession {
+
+        public SubSession(IOThread io_thread_, boolean connect_,
+                SocketBase socket_, Options options_, Address addr_) {
+            super(io_thread_, connect_, socket_, options_, addr_);
+        }
+
+    }
+    
+    public Sub(Ctx parent_, int tid_, int sid_) {
+        super(parent_, tid_, sid_);
+        
+        options.type = ZMQ.ZMQ_SUB;
+
+        //  Switch filtering messages on (as opposed to XSUB which where the
+        //  filtering is off).
+        options.filter = true;
+    }
+
+    @Override
+    public boolean xsetsockopt (int option_, Object optval_)
+    {
+        if (option_ != ZMQ.ZMQ_SUBSCRIBE && option_ != ZMQ.ZMQ_UNSUBSCRIBE) {
+            ZError.errno(ZError.EINVAL);
+            return false;
+        }
+
+        byte[] val;
+        
+        if (optval_ instanceof String)
+            val = ((String)optval_).getBytes();
+        else if (optval_ instanceof byte[])
+            val = (byte[]) optval_;
+        else
+            throw new IllegalArgumentException();
+            
+        //  Create the subscription message.
+        Msg msg = new Msg(val.length + 1);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *msg*. Currently, there are execution paths that do not contain closure statements, for example, when an if condition at line 64 evaluates to true.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).